---
title: "Targeted NET-seq Analysis"
author: "Miguel Montez"
date: "2022-12-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Packages
```{r message=FALSE, warning=FALSE}
library(dplyr)
library(data.table)
library(zoo)
library(ggh4x)
library(ggplot2)
library(ggtext)
```

# Import bedgraphs
```{r message=FALSE, warning=FALSE}
# example of loaded dataset from bedgraph
unspliced.plus.1 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.1.bedgraph", header=FALSE)
unspliced.plus.2 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.2.bedgraph", header=FALSE)
unspliced.plus.3 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.3.bedgraph", header=FALSE)
unspliced.plus.4 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.4.bedgraph", header=FALSE)
unspliced.plus.5 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.5.bedgraph", header=FALSE)
unspliced.plus.6 <- read.delim("~/R/Targeted NET-seq Analysis/unspliced.plus.6.bedgraph", header=FALSE)
```

# Label data from bedgraph and prepare dataset
```{r message=FALSE, warning=FALSE}
unsplc_reads_WT_rep1 <- unspliced.plus.1 %>% mutate(genotype="WT", bio_rep="1")
unsplc_reads_WT_rep2 <- unspliced.plus.2 %>% mutate(genotype="WT", bio_rep="2")
unsplc_reads_WT_rep3 <- unspliced.plus.3 %>% mutate(genotype="WT", bio_rep="3")
unsplc_reads_mut_rep1 <- unspliced.plus.4 %>% mutate(genotype="mut", bio_rep="1")
unsplc_reads_mut_rep2 <- unspliced.plus.5 %>% mutate(genotype="mut", bio_rep="2")
unsplc_reads_mut_rep3 <- unspliced.plus.6 %>% mutate(genotype="mut", bio_rep="3")

unsplc_reads <- rbind(unsplc_reads_WT_rep1, unsplc_reads_WT_rep2, unsplc_reads_WT_rep3, unsplc_reads_mut_rep1, unsplc_reads_mut_rep2, unsplc_reads_mut_rep3) %>% rename(chr=V1, read_end=V2, n_reads=V3)

View(unsplc_reads)

# Generate a list of all nt positions within the start and end positions of the region of interest
nt_positions_for_seq_region <- data.table(chr="5", read_end=rep(seq(18589292, 18589892), each=6), genotype=rep(c("WT", "mut"), each=3), bio_rep=c("1", "2", "3"))
nt_positions_for_seq_region <- tibble(nt_positions_for_seq_region)

# Merge the tables in order to include all positions in the NET-seq dataset
unsplc_reads_all <- merge(nt_positions_for_seq_region, unsplc_reads, by=c("chr", "read_end", "genotype", "bio_rep"), all.x=TRUE)

# Remove 5' splice sites and limit the positions for each primer to a range of 600 nt after the primer
unsplc_reads_filtered <- unsplc_reads_all %>% filter(read_end %in% 18589292:18589892)
```

# Data analysis
```{r message=TRUE, warning=TRUE}
# (1) Add pseducounts
unsplc_reads_pseudocounts <- unsplc_reads_filtered %>% mutate(n_reads=coalesce(n_reads, 0.1))

# (2) Compute the ratio

# Compute total number of reads
total_n_reads <- unsplc_reads_pseudocounts %>% group_by(genotype, bio_rep) %>% summarise(total_n_reads=sum(n_reads))

# Merge datasets to have n_reads and total_n_reads
unsplc_reads_merged <- merge(unsplc_reads_pseudocounts, total_n_reads, by=c("genotype", "bio_rep"), all.x=TRUE)

# Ratio
unsplc_reads_ratio <- unsplc_reads_merged %>% group_by(genotype, bio_rep, read_end) %>% mutate(ratio=(n_reads/total_n_reads))

# (3) Smoothing with rolling median for each bio_rep
unsplc_reads_ratio_rollmed <- unsplc_reads_ratio %>% group_by(genotype, bio_rep) %>% arrange(desc(read_end)) %>% mutate(rollmed=rollmedian(ratio, 11, fill = NA))

# (4) Average bio_rep
unsplc_reads_ratio_rollmed_avg <- unsplc_reads_ratio_rollmed %>% group_by(genotype, read_end) %>% mutate(rollmed_avg=mean(rollmed))
```

# Plot
```{r message=FALSE, warning=FALSE}
ggplot(unsplc_reads_ratio_rollmed_avg, aes(read_end, rollmed_avg, fill=genotype)) + geom_bar(stat = "identity", position = position_dodge())
```

# Prepare datasets for shading the difference between WT and mut
```{r}
unsplc_reads_ratio_rollmed_avg_WT <- unsplc_reads_ratio_rollmed_avg %>% select(genotype, read_end, rollmed_avg) %>% unique() %>% filter(genotype %in% "WT") %>% rename(WT=rollmed_avg) %>% ungroup() %>% select(-genotype)

unsplc_reads_ratio_rollmed_avg_mut <- unsplc_reads_ratio_rollmed_avg %>% select(genotype, read_end, rollmed_avg) %>% unique() %>% filter(genotype %in% "mut") %>% rename(mut=rollmed_avg) %>% ungroup() %>% select(-genotype)

shading_dataset <- merge(unsplc_reads_ratio_rollmed_avg_mut, unsplc_reads_ratio_rollmed_avg_WT, by="read_end")
```

# Plot with shading using the function stat_difference() from the ggh4x package
```{r message=FALSE, warning=FALSE}
ggplot(shading_dataset, aes(x=read_end)) + geom_line(aes(y=mut), color="#9F0032", size=0.35) + geom_line(aes(y=WT), color="#01579B", size=0.35) + stat_difference(aes(ymin=WT, ymax=mut), alpha=0.2) + scale_y_continuous(limits=c(0, 0.0042), breaks=c(0, 0.002, 0.004), expand=expansion(mult=c(0, 0.05))) + scale_x_reverse(breaks=c(18589806, 18589606, 18589406), labels=c("+1700", "+1900", "+2100"), position="top") + scale_fill_manual(values=c("#9F0032", "#01579B"), labels=c("mut", "WT")) + labs(y=NULL, x=NULL) + theme(panel.background=element_rect(fill="white"), axis.line.y.left=element_line(color="dark gray", size=0.5), axis.line.x.top=element_line(color="dark gray", size=0.5), axis.text=element_text(size=6, color="black"), axis.title.y=element_markdown(size=7, color="black"), axis.title.x=element_markdown(size=7, color="black"))
```


